### PR TITLE
Fix shard fork choice

### PR DESCRIPTION
**Pull request by hwwhww**
### Issue
The shard chain fork choice logic should be separate from different shards. (#1768 D.1)

### How did I fix it
1. fork-choice and shard-fork-choice:
    1. Create `ShardLatestMessage` dataclass to store the latest message of per shard.
    2. Add `latest_messages: Dict[ValidatorIndex, ShardLatestMessage]` field to `ShardStore`.
    3. `Store` schema update.
        - Since we have to access `ShardStore` in `on_attestation` but I also want to avoid overhaul the existing function APIs and test cases. This PR move `shard_stores: Dict[Shard, ShardStore]` to be one of the fields of `Store`.
    4. Use shard `latest_message` to calculate the shard latest attesting balance.
2. `setup.py`: handle the `dataclass` classes dependencies.

### CODE REVIEW

File: specs/phase1/shard-fork-choice.md [hunk 1]
```diff
@@ -145,9 +142,9 @@ def get_pending_shard_blocks(store: Store, shard_store: ShardStore) -> Sequence[
 #### `on_shard_block`
 
 ```python
-def on_shard_block(store: Store, shard_store: ShardStore, signed_shard_block: SignedShardBlock) -> None:
+def on_shard_block(store: Store, shard: Shard, signed_shard_block: SignedShardBlock) -> None:
```
Comment (terencechain) [inline specs/phase1/shard-fork-choice.md:145]: I wonder if we should delete `assert shard_block.shard == shard` and simplify arguments to 

`def on_shard_block(store: Store, signed_shard_block: SignedShardBlock) -> None:`

**Comment by hwwhww** [comment]
The BLS-enabled tests require #1973.

File: specs/phase1/shard-fork-choice.md (hunk 1 above)
Comment (djrtwo) [reply specs/phase1/shard-fork-choice.md:145]: agreed
We should just derive the `shard` directly from the `shard_block` and use the appropriate `shard_store` from the `store`

**Comment by djrtwo** [comment]
looks good other than merge conflict

**Comment by djrtwo** [comment]
This is still marked draft.
Any blockers for un-drafting and merging @hwwhww ?

**Comment by hwwhww** [comment]
The test would not pass if BLS flag is enabled, so kinda depending on #1980 under the surface.

**Comment by djrtwo** [comment]
#1980 merged

**Pull request merged**